### PR TITLE
Fix bug in deepcopy of utils.context.Expr

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
found a little issue with deepcopy on an `Expr` object while working on menus/contexts/keybindings ...

certainly not something that would have been a problem anywhere yet.  but will be eventually :)
This fixes and adds tests

### CODE REVIEW

File: napari/utils/context/_expressions.py [hunk 1]
```diff
@@ -315,14 +315,21 @@ def __invert__(self) -> UnaryOp[T]:
         # note: we're using the invert operator `~` to mean "not ___"
         return UnaryOp(ast.Not(), self)
 
+    def __reduce_ex__(self, protocol: int) -> None:
```
Comment (Carreau) [inline napari/utils/context/_expressions.py]:
```suggestion
    def __reduce_ex__(self, protocol: int) -> Tuple[Any, ...]:
```

File: napari/utils/context/_expressions.py [hunk 2]
```diff
@@ -315,14 +315,21 @@ def __invert__(self) -> UnaryOp[T]:
         # note: we're using the invert operator `~` to mean "not ___"
         return UnaryOp(ast.Not(), self)
 
+    def __reduce_ex__(self, protocol: int) -> Tuple[Any, ...]:
```
Comment (Carreau) [inline napari/utils/context/_expressions.py]: I think you may need:

```python
from typing import SupportsIndex
``` 

and 

```suggestion
    def __reduce_ex__(self, protocol: SupportsIndex) -> Tuple[Any, ...]:
```

Let's see how mypy complains with the current update first.

Comment (Carreau) [reply napari/utils/context/_expressions.py]: Ok, also need to import Tuple, I'll do both at the same time.

**Comment by Carreau** [comment]
Side Question, you refer to `deepcopy` in your message, but I thought `__reduce_ex__` is mostly used for pickle. I'm not sure I completely understand why, beyond maybe deepcopy using pickle to create a clone ?

**Comment by tlambert03** [comment]
> I'm not sure I completely understand why, beyond maybe deepcopy using pickle to create a clone ?

yep, relevant code here: https://github.com/python/cpython/blob/d3a27e4c9339d787f2eb3e68fd3b12e303cb4a42/Lib/copy.py#L151-L172

i used `__reduce_ex__` here after debugging because that's what the ast module itself seemed to be using.  could go higher up in the method chaing.

looks like you  fixed the typing, thanks!

**Comment by Czaki** [comment]
> Side Question, you refer to `deepcopy` in your message, but I thought `__reduce_ex__` is mostly used for pickle. I'm not sure I completely understand why, beyond maybe deepcopy using pickle to create a clone ?

if object does not provide `__deepcopy__` method then its use pickle to make deepcopy.

**Comment by Carreau** [comment]
Thanks, that's what I suspected, LGTM once test are passing.

**Review by Carreau — APPROVED** [review]

**Review by Carreau — APPROVED** [review]

**Pull request merged**